crates/ruff/src/rules/refurb/rules/repeated_append.rs

```diff
@@ -0,0 +1,377 @@
+        );
+
+        // We only suggest fix when all appends in a group are clumped together.
+        if checker.patch(diagnostic.kind.rule()) && group.is_consecutive {
+            diagnostic.set_fix(Fix::suggested(Edit::replacement(replacement, start, end)));
+        }
+
+        diagnostics.push(diagnostic);
+    }
+    checker.diagnostics.extend(diagnostics);
+}
+
+#[derive(Debug, Clone)]
+struct Append<'a> {
+    receiver: &'a ExprName,
+    binding: &'a Binding<'a>,
+    argument: &'a Expr,
+    stmt: &'a Stmt,
+}
+
+#[derive(Debug)]
+struct AppendGroup<'a> {
+    appends: Vec<Append<'a>>,
+    is_consecutive: bool,
+}
+
+impl<'a> AppendGroup<'a> {
+    fn name(&self) -> &'a str {
+        &self.appends.first().unwrap().receiver.id
+    }
+}
+
+fn match_consecutive_appends<'a>(checker: &'a Checker, stmt: &'a Stmt) -> Option<Vec<Append<'a>>> {
+    // In order to match consecutive statements, we need to go to the tree ancestor of the
+    // given statement, find its position there, and match all 'appends' from there.
+    let siblings: &'a [Stmt] = if checker.semantic().at_top_level() {
+        // If the statement is at the top level, we should go to the parent module.
+        // Module is available in the definitions list.
+        let Definition::Module(ref module_def) =
+            checker.semantic().definitions[DefinitionId::module()]
+        else {
+            return None;
+        };
+        module_def.python_ast
+    } else {
+        // Otherwise, go to the parent, and take its body as a sequence of siblings.
+        let parent = checker.semantic().current_statement_parent()?;
+        traversal::suite(stmt, parent)?
+    };
+
+    let stmt_index = siblings.iter().position(|x| x == stmt)?;
+
+    // We shouldn't repeat the same work for many 'appends' that go in a row. Let's check
+    // that this statement is at the beginning of such a group.
+    if stmt_index != 0 && match_append(checker, &siblings[stmt_index - 1]).is_some() {
+        return None;
+    }
+
+    // Starting from the current statement, let's match all appends and make a vector.
+    Some(
+        siblings
+            .iter()
+            .skip(stmt_index)
+            .map_while(|next| match_append(checker, next))
+            .collect(),
+    )
+}
+
+fn group_appends<'a>(appends: &[Append<'a>]) -> Vec<AppendGroup<'a>> {
+    // We want to go over the given list of appends and group the by receivers.
+    let mut map: HashMap<NodeId, AppendGroup<'a>> = HashMap::new();
+    let mut last_binding = {
+        let group = get_or_add(&mut map, appends.first().unwrap());
+        get_binding_stmt(group)
+    };
+
+    for append in appends.iter().skip(1) {
+        let group = get_or_add(&mut map, append);
+        let binding = get_binding_stmt(group);
+        if binding != last_binding {
+            // If the group is not brand new, and the previous group was different,
+            // we should mark it as "non-consecutive".
```

Comment:
Can you include an example here of the Python code this should catch? (I think I understand but it would be useful to see.)

Reply:
Sure!